### ISSUE TITLE
Category errors

### ISSUE THREAD

**Issue by Mostelin** [christofmuc/KnobKraft-orm#124]
Version 1.13.5: Points noted in the behaviour of Menu>Categories>Edit auto-categories:

1. Click 'Add new category': nothing happens. 'Click Save'. Reopen. Find 'New category' at bottom of list. Rename. Click 'Save'. Reopen. Find entry with new name. I think this is currently expected, but not the ultimately planned behaviour.
2. When a 22nd category is created and it becomes necessary to scroll the list, the RHS of the name border is messed up.
3. Default colour makes white text hard to read in main display. Suggestion: darker colour (perhaps background 191926). Even better: randomise new colours but constrain to muted palette. Or maybe step around colour circle in steps that are not factors of 360, e.g. 50.
4. Added categories are not recognised; any matching categories specified in JSON import file are still ignored.
5. Check mark in left column sometimes doesn't appear to update. Update can be forced by resizing the dialog box.
6. Various redrawing problems whenever dialog box is resized. This may account for points 1, 2 and 5 above.

Point 4 is the important issue. DB Browse (thanks for the tip) shows that all new categories are present and correct in the db3 file categories table. Whereas in the patches table, where it looks like you encode links to categories using bits, there are zeroes for records that should use new categories and don't fit pre-existing categories, otherwise categories are as determined by the auto-categorisation, based only on names, I believe.

**Comment by christofmuc** [comment]
@Mostelin Oh weird, in my testing this iirc worked, but yes, it sounds like a refresh/redraw problem that might hopefully be fixed.

Regarding 4) - can you give me a (truncated) PIF file so I can test in which sitation it fails?

**Comment by Mostelin** [comment]
@christofmuc 
And here is a short PIF with 5 programs that have categories 'ESQ-1', 'Hybrid', and 'Orchestral'. I deleted the last two categories from the 5th program, just in case there was some interference from categories that have not been set up. I have set up the category 'ESQ-1'.

You may have guessed that I now plan to use tags to separate ESQ-1 and SQ-80 programs, rather than operating with two adaptations. This seems the neater way to do it. However, this problem has always been there for me, it did not start with the introduction of the two new synth categories.
[Ensoniq-programs-truncated.zip](https://github.com/christofmuc/KnobKraft-orm/files/7660731/Ensoniq-programs-truncated.zip)

**Comment by christofmuc** [comment]
@Mostelin Very helpful! I found at least two bugs now in the import, this looks like in my mind had finished the work on that but actually I left where it was April 2020 and haven't returned to that function yet. But all minor and with the help of your example file I can fix it.

The refresh dynamics are a bit bigger than expected, but the general import already works. I'll prepare a new release soon!

**Comment by Mostelin** [comment]
Good to know. I am also rather relieved that this was not just me blundering around!

**Comment by christofmuc** [comment]
@Mostelin I just uploaded 1.13.6 with the bug fixes for all (?) of this mess, sorry for that. But I tested your file, and the import worked fine, just as the dialog now also behaves somewhat better.

I did not yet do better default colors - creating harmonic colors that are still distinguishable is a bit of an art, which is why I had stopped at the 11 or so categories. I like your suggestions, but for now stuck with the dark default color so people can choose their own. 

Closing this issue now in the hope of having a solution, else please reopen ;-)

**Comment by Mostelin** [comment]
Thank you for the new release 1.13.6. I have been doing a little bit of testing with it this morning and I think the GUI issues (redrawing, difficult colour combinations) are nicely resolved. And the pop-up window seems to react smoothly to resizing. Maybe the use of background grey is the best default after all.

I will try to give you some definitive information about the importing of category information later. Categories are still being ignored, but I haven't been able to sit down and figure out exactly what appears to be happening.

**Comment by Andy2No** [comment]
I haven't been able to get the auto categories I've added to do anything (in 1.14.0).

I currently have a couple of hundred Init Program patches, from importing with my Minilogue XD adaptation.  I thought I'd try separating them from the others, by auto tagging them, so I tried making a category for Init, then re-running auto categorisation, while filtering for "Init", or "Untagged", or both.  Nothing happens.

I also tried making an auto category for "Prog" - same result.

I thought maybe programs called "Init Program" are excluded, and I also have a couple of dozen factory patches starting with "TPL...", which are as yet untagged, so I thought I'd try making a TPL category - that doesn't work either.

Am I just not understanding what to do?

Incidentally, when I tried the **Categories -> Edit category import mapping** menu item, I just got an error saying the mapping_categories.jsonc file doesn't exist, with a couple of options on what to do, apparently offered by Windows, in a dialogue box.

So, based on what I've tried so far, I think this issue may have been closed before its time.
